eventauth.go

```diff
@@ -859,35 +857,37 @@ func (e *eventAllower) commonChecks(event PDU) error {
 		)
 	}
 
-	sender := event.Sender()
 	stateKey := event.StateKey()
-
-	if err := e.create.UserIDAllowed(sender); err != nil {
+	userID, err := event.UserID()
+	if err != nil {
+		return err
+	}
+	if err := e.create.UserIDAllowed(userID); err != nil {
 		return err
 	}
 
 	// Check that the sender is in the room.
 	// Every event other than m.room.create, m.room.member and m.room.aliases require this.
 	if e.member.Membership != spec.Join {
-		return errorf("sender %q not in room", sender)
+		return errorf("sender %q not in room", event.SenderID())
 	}
 
-	senderLevel := e.powerLevels.UserLevel(sender)
+	senderLevel := e.powerLevels.UserLevel(event.SenderID())
 	eventLevel := e.powerLevels.EventLevel(event.Type(), stateKey != nil)
 	if senderLevel < eventLevel {
 		return errorf(
 			"sender %q is not allowed to send event. %d < %d",
-			event.Sender(), senderLevel, eventLevel,
+			event.SenderID(), senderLevel, eventLevel,
 		)
 	}
 
 	// Check that all state_keys that begin with '@' are only updated by users
 	// with that ID.
 	if stateKey != nil && len(*stateKey) > 0 && (*stateKey)[0] == '@' {
-		if *stateKey != sender {
+		if *stateKey != event.SenderID() {
```

Comment:
This check is what I've spoken about before, which is a problem for pseudo IDs as they do NOT have `@` sigil, and yet will be used in cases where the events are only intended to be updated by users with the same ID. I don't think this is a strong reason to introduce a new sigil e.g `^` like MSC1228 suggests. We do however need to figure out a way to resolve this.